shared/strbuf.c

```diff
@@ -8,75 +8,111 @@
 #include <stdbool.h>
 #include <stdlib.h>
 #include <string.h>
+#include <sys/param.h>
 
 #include "util.h"
 #include "strbuf.h"
 
 #define BUF_STEP 128
 
-static bool buf_grow(struct strbuf *buf, size_t newsize)
+static bool buf_realloc(struct strbuf *buf, size_t sz)
 {
 	void *tmp;
-	size_t sz;
 
-	if (newsize <= buf->size)
-		return true;
-
-	if (newsize % BUF_STEP == 0)
-		sz = newsize;
+	if (!buf->heap)
+		tmp = malloc(sz);
 	else
-		sz = ((newsize / BUF_STEP) + 1) * BUF_STEP;
+		tmp = realloc(buf->bytes, sz);
 
-	tmp = realloc(buf->bytes, sz);
-	if (sz > 0 && tmp == NULL)
-		return false;
+	if (sz > 0) {
+		if (tmp == NULL)
+			return false;
+
+		if (!buf->heap)
+			memcpy(tmp, buf->bytes, MIN(buf->size, sz));
+	}
+
+	buf->heap = true;
 	buf->bytes = tmp;
 	buf->size = sz;
+
 	return true;
 }
 
+bool strbuf_reserve(struct strbuf *buf, size_t n)
+{
+	n += buf->used;
```

Comment:
We could use `uaddsz_overflow` here to make sure to never overflow, no matter how huge function argument `n`.

Reply:
done